core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1746,4 +1779,17 @@ class ReplicaManager(val config: KafkaConfig,
 
     controller.electLeaders(partitions, electionType, electionCallback)
   }
+
+  class FollowerLagStats {
+    private val lagStats = new ConcurrentHashMap[TopicPartition, Long]()
```

Comment:
After reviewing this patch, I'm starting to wonder if we should be looking at a different kind of metric. The annoying thing here is the need to track all this extra state as well as the need to scan all values to find the current max. I can't think of a great solution unfortunately. Maybe one way to do it is to let these gauges reference an atomic value which is updated periodically by a background task. Then we can compute several metric values by the background task and keep performance bounded.

Reply:
So the problem that we're trying to solve is to provide an easy way for monitoring systems to detect (and perhaps mitigate) stalling reassignment. I think on this level it may just be better to introduce partition or even replica level reassignment lag metrics and let the monitoring systems define how do they detect stalling reassignment and perhaps mitigate it.
My reasons for this are the following:
a) Kafka won't do anything with the reassignment lag so if it's too ineffective to compute for us, maybe we shouldn't. Usually monitoring runs on remote hosts so we can decouple the problem.
b) From a monitoring or management system's perspective it's much better if we provide more fine-grained metrics as they can build up more robust strategies from that.
c) It is much easier to create a Meter for every partition and update that. In this case we might not even need to track the lag on the follower side neither as we'll be able to tell which broker is not replicating by looking at the leader side partition level metrics. Also internalizing it in Partition or Replica would make the metric's life cycle equal to that so we don't have to maintain it separately.
d) If we update something periodically we will need to figure out a good period which caters most of the use cases. Monitoring systems can use varying periods and we want to cater them all and it can be hard to stay computation friendly yet support the variety of monitoring system configurations (you can scan once per second or once per minute too).

What do you think?

Reply:
(Having said this I'm also good with defining all this in a subsequent KIP if you agree and think this is to big of a piece for now.)

Reply:
It's kind of a double-edged sword with partition-level metrics. They are easier to update, but they are more expensive to collect and store. How about we take the lag changes out of this patch and do them separately? I don't think we need a new KIP, but I do think they need a little more consideration and I'm hoping we can get fixed URP logic checked in soon.

Reply:
Sure, I agree with that. I'll take them out.

Reply:
Agreed with all said.

I believe we should aim for metrics that have the least cardinality. Having many per-partition metrics puts a lot of strain on monitoring systems. 

For example, the [datadog agent](https://docs.datadoghq.com/agent/), has to periodically search all matching beans every X minutes in order to detect new ones. The reason, I believe, is because JMX has a limitation where you can't query individual attributes without knowing their types - and to know the types you need to have all the attributes.
JMX has some filtering capabilities but they are very limited as far as I know. (cc @xvrl in case he has any thoughts on the matter)

This becomes an even bigger problem once we make Kafka support more partitions. Such public changes aren't easy to reverse once we find them problematic